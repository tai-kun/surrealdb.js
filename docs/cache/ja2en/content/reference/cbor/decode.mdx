---
title: decode
---

## Synchronous Decoding

The `decode` function decodes a CBOR-formatted `Uint8Array` into a JavaScript value.

### Import

```ts
import { decode } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decode(input: Uint8Array, options?: DecodeOptions): unknown;
```

#### Arguments

##### `input`

The CBOR-formatted `Uint8Array` to decode into a JavaScript value.

##### `options`

Decoding options.

`mapType?: "Object" | "Map"`
:   This option can be used to choose whether to decode key-value objects as JavaScript plain objects or `Map` objects. Choose `"Object"` to decode as a plain object and `"Map"` to decode as a `Map` class instance. The default is `"Object"`. Plain object keys can only be strings or numbers (i.e., `string | number`), but if you decode with this option set to `"Map"`, the keys can be any decoded value.

`reviver?: Reviver | ReviverObject | (Reviver | ReviverObject)[]`
:   This option is used to transform CBOR simple data items or tagged data items into JavaScript values. Simple data items are limited to values from 0 to 19 and 32 to 255. The tag of a tagged data item is a `number` or `bigint` from 0 to 2^64-1. If the function returns the `CONTINUE` symbol, which can be obtained with `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions will be called. If the value cannot be transformed, `SurrealTypeError` is thrown. 

`isSafeMapKey?: (key: unknown, map: Map<unknown, unknown>) => boolean`
:   This option is a function that validates whether a key of a `Map` object is safe. The function takes two arguments: `key` and `map`. `key` is the value of the key, and `map` is the `Map` object being validated. If this function returns `false`, `CborUnsafeMapKeyError` is thrown. The default considers `"__proto__"` or `"constructor"` as unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: Record<string | number, unknown>) => boolean`
:   This option is a function that validates whether a key of a plain object is safe. The function takes two arguments: `key` and `obj`. `key` is the value of the key, and `obj` is the plain object being validated. If this function returns `false`, `CborUnsafeMapKeyError` is thrown. The default considers `"__proto__"` or `"constructor"` as unsafe keys. Note that specifying this option will completely override the default validation. If you want to add your own validation to the default validation, you should also consider `"__proto__"` or `"constructor"` as unsafe keys.

`maxDepth?: number`
:   Specifies the maximum depth of the object. The depth is incremented by 1 for each object or array entered. The default value is `64`. Since recursion is not used for decoding, there is no upper limit to the depth, but it is recommended to keep it small.

#### Return Value

The JavaScript value decoded from the CBOR-formatted `Uint8Array`.

### Example

#### Decoding as a `Map` object

```ts
import { decode } from "@tai-kun/surrealdb/cbor";

const cbor = new Uint8Array([
  0xa1, // mt: 5, ai: 1
  // key
  0x00, // mt: 0, ai: 0
  // value
  0x00, // mt: 0, ai: 0
]);

const value = decode(bytes, { mapType: "Map" });

console.log(value); // Map(1) { [ 0 ] => 0 }
```

## Asynchronous Decoding

The `decodeStream` function decodes a CBOR-formatted `Uint8Array` into a JavaScript value.

### Import

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decodeStream(
  input: ReadableStream<Uint8Array>,
  options?: DecodeStreamOptions,
): unknown;
```

#### Arguments

##### `input`

A readable stream of `Uint8Array`. The Fetch API may provide this stream on the `body` property of the `Response` object in response to a request.

##### `options`

In addition to the options for synchronous decoding, the following options can be specified:

`signal?: AbortSignal`
:   An AbortSignal to abort the stream.

#### Return Value

A `StatefulPromise` that resolves with the JavaScript value.

### Example

#### Fetch API

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";

const response = await fetch("http://localhost:8000/rpc", {
  method: "POST",
  headers: {
    Accept: "application/cbor",
    // ... more headers
  },
  // ... more properties
})

const result = await decodeStream(resp.body);
```
