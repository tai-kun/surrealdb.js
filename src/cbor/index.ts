export type * from "./decode-stream";
export { default as decodeStream } from "./decode-stream";

export type * from "./decode";
export { default as decode } from "./decode";

export {
  CONTINUE,
  type IsSafeMapKey,
  type IsSafeObjectKey,
  type Reviver,
  type ReviverObject,
  type SimpleItemReviver,
  type TaggedItemReviver,
} from "./decoder";

export type * from "./encode";
export { default as encode } from "./encode";

export {
  type AdditionalInfo,
  AI_EIGHT_BYTES,
  AI_FOUR_BYTES,
  AI_INDEFINITE_NUM_BYTES,
  AI_ONE_BYTE,
  AI_SIMPLE_FALSE,
  AI_SIMPLE_NULL,
  AI_SIMPLE_TRUE,
  AI_SIMPLE_UNDEFINED,
  AI_TWO_BYTES,
  type DataItem,
  type DataItemValue,
  HEADER_FALSE,
  HEADER_FLOAT_DOUBLE,
  HEADER_FLOAT_HALF,
  HEADER_NULL,
  HEADER_TRUE,
  HEADER_UNDEFINED,
  type MajorType,
  MT_ARRAY,
  MT_BYTE_STRING,
  MT_MAP,
  MT_NEGATIVE_INTEGER,
  MT_SIMPLE_FLOAT,
  MT_TAG,
  MT_UNSIGNED_INTEGER,
  MT_UTF8_STRING,
  Simple,
} from "./spec";

export type * from "./tagged";
export { default as Tagged } from "./tagged";

export type { ToCBOR } from "./traits";
export { canToCBOR } from "./traits";

export {
  type Replacer,
  write,
  writeBigInt,
  writeBoolean,
  writeByteString,
  writeEncodedUtf8String,
  writeHeader,
  writeNullable,
  writeNumber,
  writePayload,
  writeUtf8String,
} from "./write-utils";

export type * from "./writer"; // `Writer` も型だけでよし
