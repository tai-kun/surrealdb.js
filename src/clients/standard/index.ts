export type * from "./client";
export { default } from "./client";

export type * from "./jwt";
export { default as Jwt } from "./jwt";
