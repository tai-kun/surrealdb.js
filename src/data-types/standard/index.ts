export type * from "./src/datetime";
export { default as Datetime } from "./src/datetime";

export type * from "./src/decimal";
export { default as Decimal } from "./src/decimal";

export type * from "./src/duration";
export { default as Duration } from "./src/duration";

export {
  type GeoJsonCollection,
  GeometryCollection,
  GeometryCollectionBase,
  type GeometryCollectionSource,
  type GeometryCollectionTypes,
} from "./src/geometry-collection";
export {
  type GeoJsonLineString,
  GeometryLine,
  GeometryLineBase,
  type GeometryLineSource,
  type GeometryLineTypes,
} from "./src/geometry-line";
export {
  type GeoJsonMultiLine,
  GeometryMultiLine,
  GeometryMultiLineBase,
  type GeometryMultiLineSource,
  type GeometryMultiLineTypes,
} from "./src/geometry-multiline";
export {
  type GeoJsonMultiPoint,
  GeometryMultiPoint,
  GeometryMultiPointBase,
  type GeometryMultiPointSource,
  type GeometryMultiPointTypes,
} from "./src/geometry-multipoint";
export {
  type GeoJsonMultiPolygon,
  GeometryMultiPolygon,
  GeometryMultiPolygonBase,
  type GeometryMultiPolygonSource,
  type GeometryMultiPolygonTypes,
} from "./src/geometry-multipolygon";
export {
  type GeoJsonPoint,
  GeometryPoint,
  GeometryPointBase,
  type GeometryPointSource,
  type GeometryPointTypes,
} from "./src/geometry-point";
export {
  type GeoJsonPolygon,
  GeometryPolygon,
  GeometryPolygonBase,
  type GeometryPolygonSource,
  type GeometryPolygonTypes,
} from "./src/geometry-polygon";

export type * from "./src/table";
export { default as Table } from "./src/table";

export type * from "./src/thing";
export { default as Thing } from "./src/thing";

export type * from "./src/uuid";
export { default as Uuid } from "./src/uuid";

export {
  CBOR_TAG_CUSTOM_DATETIME,
  CBOR_TAG_CUSTOM_DURATION,
  CBOR_TAG_GEOMETRY_COLLECTION,
  CBOR_TAG_GEOMETRY_LINE,
  CBOR_TAG_GEOMETRY_MULTILINE,
  CBOR_TAG_GEOMETRY_MULTIPOINT,
  CBOR_TAG_GEOMETRY_MULTIPOLYGON,
  CBOR_TAG_GEOMETRY_POINT,
  CBOR_TAG_GEOMETRY_POLYGON,
  CBOR_TAG_NONE,
  CBOR_TAG_RECORDID,
  CBOR_TAG_SPEC_DATETIME,
  CBOR_TAG_SPEC_UUID,
  CBOR_TAG_STRING_DECIMAL,
  CBOR_TAG_STRING_DURATION,
  CBOR_TAG_STRING_UUID,
  CBOR_TAG_TABLE,
  type Encodable,
} from "../encodable/src/spec";
