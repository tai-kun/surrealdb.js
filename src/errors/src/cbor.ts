import { SurrealError, type SurrealErrorOptions } from "./general";

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborerror)
 */
export class CborError extends SurrealError {
  static {
    this.prototype.name = "CborError";
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#name-well-formedness-errors-and-
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborwellformednesserror)
 */
export class CborWellFormednessError extends CborError {
  static {
    this.prototype.name = "CborWellFormednessError";
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.2
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbortoomuchdataerror)
 */
export class CborTooMuchDataError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborTooMuchDataError";
  }

  constructor(options?: SurrealErrorOptions | undefined) {
    super("Unconsumed input bytes remain after decoding.", options);
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.4
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbortoolittledataerror)
 */
export class CborTooLittleDataError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborTooLittleDataError";
  }

  constructor(options?: SurrealErrorOptions | undefined) {
    super(
      "Input data appears truncated or incomplete for CBOR decoding.",
      options,
    );
  }
}

// https://datatracker.ietf.org/doc/html/rfc8949#section-appendix.f-2.6
/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborsyntaxerror)
 */
export class CborSyntaxError extends CborWellFormednessError {
  static {
    this.prototype.name = "CborSyntaxError";
  }
}

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cbormaxdepthreachederror)
 */
export class CborMaxDepthReachedError extends CborError {
  static {
    this.prototype.name = "CborMaxDepthReachedError";
  }

  constructor(
    public maxDepth: number,
    options?: SurrealErrorOptions | undefined,
  ) {
    super(`Maximum depth of ${maxDepth} has been reached.`, options);
  }
}

/**
 * [API Reference](https://tai-kun.github.io/surrealdb.js/reference/errors/#cborunsafemapkeyerror)
 */
export class CborUnsafeMapKeyError extends CborError {
  static {
    this.prototype.name = "CborUnsafeMapKeyError";
  }

  constructor(
    public key: unknown,
    options?: SurrealErrorOptions | undefined,
  ) {
    super(
      `Invalid key for CBOR map: ${String(key)}. The key must be a valid CBOR`
        + " data type that is safe to use in a JavaScript map or object.",
      options,
    );
  }
}

export class CborDecodeStreamAbortFailedError extends CborError {
  static {
    this.prototype.name = "CborStreamAbortFailedError";
  }

  override cause: unknown[];

  constructor(
    errors: readonly unknown[],
    options?: Omit<SurrealErrorOptions, "cause"> | undefined,
  ) {
    super("Failed to abort decode-stream", options);
    this.cause = errors.slice();
  }
}

export class CborMemoryError extends CborError {
  static {
    this.prototype.name = "CborMemoryError";
  }
}

export class CborMemoryBlockError extends CborError {
  static {
    this.prototype.name = "CborMemoryBlockError";
  }
}

type MemoryAddress = string | number;

export class CborMemoryBlockConflictError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborMemoryBlockConflictError";
  }

  constructor(
    address: MemoryAddress,
    expectedSize: number,
    actualSize: number,
  ) {
    super(
      `Memory conflict at address ${address}: `
        + `Expected size ${expectedSize}, but found ${actualSize}.`,
    );
  }
}

export class CborMemoryBlockInUseError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborMemoryBlockInUseError";
  }
}

export class CborUndefinedMemoryBlockError extends CborMemoryBlockError {
  static {
    this.prototype.name = "CborUndefinedMemoryBlockError";
  }

  constructor(address: MemoryAddress) {
    super(`Memory block at address ${address} is not defined.`);
  }
}
