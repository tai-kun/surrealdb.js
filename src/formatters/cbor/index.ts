export type * from "./formatter";
export { default } from "./formatter";
