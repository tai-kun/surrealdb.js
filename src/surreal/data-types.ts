import type {
  Datetime as DecodeOnlyDatetime,
  Decimal as DecodeOnlyDecimal,
  Duration as DecodeOnlyDuration,
  GeometryCollection as DecodeOnlyGeometryCollection,
  GeometryLine as DecodeOnlyGeometryLine,
  GeometryMultiLine as DecodeOnlyGeometryMultiLine,
  GeometryMultiPoint as DecodeOnlyGeometryMultiPoint,
  GeometryMultiPolygon as DecodeOnlyGeometryMultiPolygon,
  GeometryPoint as DecodeOnlyGeometryPoint,
  GeometryPolygon as DecodeOnlyGeometryPolygon,
  Table as DecodeOnlyTable,
  Thing as DecodeOnlyThing,
  Uuid as DecodeOnlyUuid,
  // Future as DecodeOnlyFuture,
  // Range as DecodeOnlyRange,
  // BoundIncluded as DecodeOnlyBoundIncluded,
  // BoundExcluded as DecodeOnlyBoundExcluded,
} from "@tai-kun/surrealdb/data-types/decode-only";
import type {
  Datetime as EncodableDatetime,
  Decimal as EncodableDecimal,
  Duration as EncodableDuration,
  GeometryCollection as EncodableGeometryCollection,
  GeometryLine as EncodableGeometryLine,
  GeometryMultiLine as EncodableGeometryMultiLine,
  GeometryMultiPoint as EncodableGeometryMultiPoint,
  GeometryMultiPolygon as EncodableGeometryMultiPolygon,
  GeometryPoint as EncodableGeometryPoint,
  GeometryPolygon as EncodableGeometryPolygon,
  Table as EncodableTable,
  Thing as EncodableThing,
  Uuid as EncodableUuid,
  // Future as EncodableFuture,
  // Range as EncodableRange,
  // BoundIncluded as EncodableBoundIncluded,
  // BoundExcluded as EncodableBoundExcluded,
} from "@tai-kun/surrealdb/data-types/encodable";
import type {
  Datetime as StandardDatetime,
  Decimal as StandardDecimal,
  Duration as StandardDuration,
  GeometryCollection as StandardGeometryCollection,
  GeometryLine as StandardGeometryLine,
  GeometryMultiLine as StandardGeometryMultiLine,
  GeometryMultiPoint as StandardGeometryMultiPoint,
  GeometryMultiPolygon as StandardGeometryMultiPolygon,
  GeometryPoint as StandardGeometryPoint,
  GeometryPolygon as StandardGeometryPolygon,
  Table as StandardTable,
  Thing as StandardThing,
  Uuid as StandardUuid,
  // Future as StandardFuture,
  // Range as StandardRange,
  // BoundIncluded as StandardBoundIncluded,
  // BoundExcluded as StandardBoundExcluded,
} from "@tai-kun/surrealdb/data-types/standard";

export namespace DataType {
  export type Table =
    | DecodeOnlyTable
    | EncodableTable
    | StandardTable;

  export type Thing =
    | DecodeOnlyThing
    | EncodableThing
    | StandardThing;

  export type Decimal =
    | DecodeOnlyDecimal
    | EncodableDecimal
    | StandardDecimal;

  export type Datetime =
    | DecodeOnlyDatetime
    | EncodableDatetime
    | StandardDatetime;

  export type Duration =
    | DecodeOnlyDuration
    | EncodableDuration
    | StandardDuration;

  // export type Future =
  //   | DecodeOnlyFuture
  //   | EncodableFuture
  //   | StandardFuture;

  export type Uuid =
    | DecodeOnlyUuid
    | EncodableUuid
    | StandardUuid;

  // export type Range =
  //   | DecodeOnlyRange
  //   | EncodableRange
  //   | StandardRange;

  // export type BoundIncluded =
  //   | DecodeOnlyBoundIncluded
  //   | EncodableBoundIncluded
  //   | StandardBoundIncluded;

  // export type BoundExcluded =
  //   | DecodeOnlyBoundExcluded
  //   | EncodableBoundExcluded
  //   | StandardBoundExcluded;

  export type GeometryPoint =
    | DecodeOnlyGeometryPoint
    | EncodableGeometryPoint
    | StandardGeometryPoint;

  export type GeometryLine =
    | DecodeOnlyGeometryLine
    | EncodableGeometryLine
    | StandardGeometryLine;

  export type GeometryPolygon =
    | DecodeOnlyGeometryPolygon
    | EncodableGeometryPolygon
    | StandardGeometryPolygon;

  export type GeometryMultiPoint =
    | DecodeOnlyGeometryMultiPoint
    | EncodableGeometryMultiPoint
    | StandardGeometryMultiPoint;

  export type GeometryMultiLine =
    | DecodeOnlyGeometryMultiLine
    | EncodableGeometryMultiLine
    | StandardGeometryMultiLine;

  export type GeometryMultiPolygon =
    | DecodeOnlyGeometryMultiPolygon
    | EncodableGeometryMultiPolygon
    | StandardGeometryMultiPolygon;

  export type GeometryCollection =
    | DecodeOnlyGeometryCollection
    | EncodableGeometryCollection
    | StandardGeometryCollection;
}

export type DataType =
  | DataType.Table
  | DataType.Thing
  | DataType.Decimal
  | DataType.Datetime
  | DataType.Duration
  // | DataType.Future
  | DataType.Uuid
  // | DataType.Range
  // | DataType.BoundIncluded
  // | DataType.BoundExcluded
  | DataType.GeometryPoint
  | DataType.GeometryLine
  | DataType.GeometryPolygon
  | DataType.GeometryMultiPoint
  | DataType.GeometryMultiLine
  | DataType.GeometryMultiPolygon
  | DataType.GeometryCollection;

function isValue(o: unknown, id: string): boolean {
  return !!o
    && typeof o === "object"
    // @ts-expect-error
    && o["$$datatype"] === Symbol.for("@tai-kun/surrealdb/data-types/" + id);
}

export function isTable<T = DataType.Table>(o: unknown): o is T {
  return isValue(o, "table");
}

export function isThing<T = DataType.Thing>(o: unknown): o is T {
  return isValue(o, "thing");
}

export function isDecimal<T = DataType.Decimal>(o: unknown): o is T {
  return isValue(o, "decimal");
}

export function isDatetime<T = DataType.Datetime>(o: unknown): o is T {
  return isValue(o, "datetime");
}

export function isDuration<T = DataType.Duration>(o: unknown): o is T {
  return isValue(o, "duration");
}

// export function isFuture<T = DataType.Future>(o: unknown): o is T {
//   return isValue(o, "future");
// }

export function isUuid<T = DataType.Uuid>(o: unknown): o is T {
  return isValue(o, "uuid");
}

// export function isRange<T = DataType.Range>(o: unknown): o is T {
//   return isValue(o, "range");
// }

// export function isBoundIncluded<T = DataType.BoundIncluded>(
//   o: unknown,
// ): o is T {
//   return isValue(o, "boundincluded");
// }

// export function isBoundExcluded<T = DataType.BoundExcluded>(
//   o: unknown,
// ): o is T {
//   return isValue(o, "boundexcluded");
// }

export function isGeometryPoint<T = DataType.GeometryPoint>(
  o: unknown,
): o is T {
  return isValue(o, "geometrypoint");
}

export function isGeometryLine<T = DataType.GeometryLine>(o: unknown): o is T {
  return isValue(o, "geometryline");
}

export function isGeometryPolygon<T = DataType.GeometryPolygon>(
  o: unknown,
): o is T {
  return isValue(o, "geometrypolygon");
}

export function isGeometryMultiPoint<T = DataType.GeometryMultiPoint>(
  o: unknown,
): o is T {
  return isValue(o, "geometrymultipoint");
}

export function isGeometryMultiLine<T = DataType.GeometryMultiLine>(
  o: unknown,
): o is T {
  return isValue(o, "geometrymultiline");
}

export function isGeometryMultiPolygon<T = DataType.GeometryMultiPolygon>(
  o: unknown,
): o is T {
  return isValue(o, "geometrymultipolygon");
}

export function isGeometryCollection<T = DataType.GeometryCollection>(
  o: unknown,
): o is T {
  return isValue(o, "geometrycollection");
}
