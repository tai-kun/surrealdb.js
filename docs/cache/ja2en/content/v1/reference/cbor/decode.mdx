---
title: Decode
slug: en/v1/reference/cbor/decode
---

## Synchronous Decoding

The `decode` function decodes a `Uint8Array` in CBOR format into a JavaScript value.

### Import

```ts
import { decode } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decode(input: Uint8Array, options?: DecodeOptions): unknown;
```

#### Arguments

##### `input`

A `Uint8Array` in CBOR format to be decoded into a JavaScript value.

##### `options`

Decoding options.

`mapType?: "Object" | "Map"`
:   This option can be used to choose whether to decode key-value objects as JavaScript plain objects or `Map` objects. Selecting `"Object"` decodes as a plain object, while `"Map"` decodes as an instance of the `Map` class. The default value is `"Object"`. Plain object keys can only accept strings or numbers (i.e. `string | number`), but decoding with this option set to `"Map"` allows keys to accept all decoded values.

`reviver?: Reviver | ReviverObject | (Reviver | ReviverObject)[]`
:   This option is used to convert simple or tagged data items from CBOR to JavaScript values. Simple data items are limited to values between 0 and 19 and between 32 and 255. The tag of a tagged data item is a `number` or `bigint` between 0 and 2^64-1. If the function returns the `CONTINUE` symbol obtained with `Symbol.for("@tai-kun/surrealdb/cbor/continue")`, subsequent functions are called. If no conversion could be made, a `SurrealTypeError` is thrown.

`isSafeMapKey?: (key: unknown, map: Map<unknown, unknown>) => boolean`
:   This option is a function that verifies whether a key in a `Map` object is safe. The function accepts two arguments: `key`, which is the key value, and `map`, which is the `Map` object being validated. If this function returns `false`, a `CborUnsafeMapKeyError` is thrown. By default, it determines `"__proto__"` or `"constructor"` as unsafe keys. Please note that specifying this option completely overrides the default validation. If you want to add your own validation to the default validation, you should also determine `"__proto__"` or `"constructor"` as unsafe keys.

`isSafeObjectKey?: (key: unknown, obj: Record<string | number, unknown>) => boolean`
:   This option is a function that verifies whether a key in a plain object is safe. The function accepts two arguments: `key`, which is the key value, and `obj`, which is the plain object being validated. If this function returns `false`, a `CborUnsafeMapKeyError` is thrown. By default, it determines `"__proto__"` or `"constructor"` as unsafe keys. Please note that specifying this option completely overrides the default validation. If you want to add your own validation to the default validation, you should also determine `"__proto__"` or `"constructor"` as unsafe keys.

`maxDepth?: number`
:   Specifies the maximum depth of objects. The depth increases by 1 for every object or array entry. The default value is `64`. While decoding does not use recursive processing, there is no upper limit on depth, but it is recommended that it not be too large.

#### Return value

The JavaScript value decoded from the `Uint8Array` in CBOR format.

### Example

#### Decoding with `Map` objects

```ts
import { decode } from "@tai-kun/surrealdb/cbor";

const cbor = new Uint8Array([
  0xa1, // mt: 5, ai: 1
  // key
  0x00, // mt: 0, ai: 0
  // value
  0x00, // mt: 0, ai: 0
]);

const value = decode(bytes, { mapType: "Map" });

console.log(value); // Map(1) { [ 0 ] => 0 }
```

## Asynchronous Decoding

The `decodeStream` function decodes a `Uint8Array` in CBOR format into a JavaScript value.

### Import

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";
```

### Syntax

```ts
function decodeStream(
  input: ReadableStream<Uint8Array>,
  options?: DecodeStreamOptions,
): unknown;
```

#### Arguments

##### `input`

A readable stream of `Uint8Array`. The Fetch API may provide this stream in the `body` property of the `Response` object for requests.

##### `options`

In addition to the options for synchronous decoding, the following options can be specified:

`signal?: AbortSignal`
:   An abort signal to interrupt the stream.

#### Return value

A `StatefulPromise` that resolves to a JavaScript value.

### Example

#### Fetch API

```ts
import { decodeStream } from "@tai-kun/surrealdb/cbor";

const response = await fetch("http://localhost:8000/rpc", {
  method: "POST",
  headers: {
    Accept: "application/cbor",
    // ... more headers
  },
  // ... more properties
})

const result = await decodeStream(resp.body);
```
