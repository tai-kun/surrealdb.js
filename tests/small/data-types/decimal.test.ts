import { isDecimal } from "@tai-kun/surrealdb";
import { decode, encode } from "@tai-kun/surrealdb/cbor";
import {
  CBOR_TAG_STRING_DECIMAL,
  Decimal as EncodableDecimal,
} from "@tai-kun/surrealdb/data-types/encodable";
import { Decimal } from "@tai-kun/surrealdb/data-types/standard";
import { describe, expect, test } from "vitest";

// -----------------------------------------------------------------------------
//
// Decode-only / Encodable Decimal
//
// -----------------------------------------------------------------------------

describe("decode-only/encodable", () => {
  describe("有効な文字列", () => {
    test(".toString()", () => {
      expect((new EncodableDecimal("1.23e1")).toString()).toBe("1.23e1");
    });

    test("テンプレートリテラル", () => {
      expect(`${new EncodableDecimal("1.23e1")}`).toBe("1.23e1");
    });

    test("文字列へ暗黙の型変換", () => {
      expect("" + new EncodableDecimal("1.23e1")).toBe("1.23e1");
    });

    test("数値へ暗黙の型変換", () => {
      expect(+new EncodableDecimal("1.23e1")).toBe(12.3);
    });

    test("CBOR でエンコード/デコードできる", () => {
      const input = new EncodableDecimal("1.23e1");
      const output = new EncodableDecimal("1.23e1");
      const bytes = encode(input);
      const dt = decode(bytes, {
        reviver: {
          tagged(t) {
            switch (t.tag) {
              case CBOR_TAG_STRING_DECIMAL:
                return new EncodableDecimal(t.value as any);

              default:
                return undefined;
            }
          },
        },
      });

      expect(dt).toStrictEqual(output);
    });

    test(".toJSON()", () => {
      const json = new EncodableDecimal("1.23e1").toJSON();

      expect(json).toBe("1.23e1");
    });

    test(".toSurql()", () => {
      const surql = new EncodableDecimal("1.23e1").toSurql();

      expect(surql).toBe("1.23e1dec");
    });

    test(".toPlain()", () => {
      const toPlain = new EncodableDecimal("1.23e1").toPlain();

      expect(toPlain).toStrictEqual({
        value: "1.23e1",
      });
    });
  });

  describe("無効な文字列", () => {
    test("インスタンスの作成に失敗する", () => {
      expect(() => new EncodableDecimal("+3.14"))
        .toThrowErrorMatchingSnapshot();
    });
  });
});

// -----------------------------------------------------------------------------
//
// Standard Decimal
//
// -----------------------------------------------------------------------------

describe("standard", () => {
  describe("有効な文字列", () => {
    test(".toString()", () => {
      expect((new Decimal("1.23e1")).toString()).toBe("12.3");
    });

    test("テンプレートリテラル", () => {
      expect(`${new Decimal("1.23e1")}`).toBe("12.3");
    });

    test("文字列へ暗黙の型変換", () => {
      expect("" + new Decimal("1.23e1")).toBe("12.3");
    });

    test("数値へ暗黙の型変換", () => {
      expect(+new Decimal("1.23e1")).toBe(12.3);
    });

    test("CBOR でエンコード/デコードできる", () => {
      const input = new Decimal("1.23e1");
      const output = new Decimal("1.23e1");
      const bytes = encode(input);
      const dt = decode(bytes, {
        reviver: {
          tagged(t) {
            switch (t.tag) {
              case CBOR_TAG_STRING_DECIMAL:
                return new Decimal(t.value as any);

              default:
                return undefined;
            }
          },
        },
      });

      expect(dt).toStrictEqual(output);
    });

    test(".toJSON()", () => {
      const json = new Decimal("1.23e1").toJSON();

      expect(json).toBe("12.3");
    });

    test(".toSurql()", () => {
      const surql = new Decimal("1.23e1").toSurql();

      expect(surql).toBe("12.3dec");
    });

    test(".toPlain()", () => {
      const toPlain = new Decimal("1.23e1").toPlain();

      expect(toPlain).toStrictEqual({
        value: "12.3",
        exponent: 1,
        sign: 1,
        singleDigits: [
          1,
          2,
          3,
        ],
      });
    });
  });

  describe("無効な文字列", () => {
    test("インスタンスの作成に失敗する", () => {
      expect(() => new Decimal("+3.14"))
        .toThrowErrorMatchingSnapshot();
    });
  });

  describe("有効な数値", () => {
    test(".toString()", () => {
      expect((new Decimal(3.14)).toString()).toBe("3.14");
    });

    test("テンプレートリテラル", () => {
      expect(`${new Decimal(3.14)}`).toBe("3.14");
    });

    test("文字列へ暗黙の型変換", () => {
      expect("" + new Decimal(3.14)).toBe("3.14");
    });

    test("数値へ暗黙の型変換", () => {
      expect(+new Decimal(3.14)).toBe(3.14);
    });

    test("CBOR でエンコード/デコードできる", () => {
      const input = new Decimal(3.14);
      const output = new Decimal(3.14);
      const bytes = encode(input);
      const dt = decode(bytes, {
        reviver: {
          tagged(t) {
            switch (t.tag) {
              case CBOR_TAG_STRING_DECIMAL:
                return new Decimal(t.value as any);

              default:
                return undefined;
            }
          },
        },
      });

      expect(dt).toStrictEqual(output);
    });

    test(".toJSON()", () => {
      const json = new Decimal(3.14).toJSON();

      expect(json).toBe("3.14");
    });

    test(".toSurql()", () => {
      const surql = new Decimal(3.14).toSurql();

      expect(surql).toBe("3.14dec");
    });

    test(".toPlain()", () => {
      const toPlain = new Decimal(3.14).toPlain();

      expect(toPlain).toStrictEqual({
        value: "3.14",
        exponent: 0,
        sign: 1,
        singleDigits: [
          3,
          1,
          4,
        ],
      });
    });
  });

  describe("無効な数値", () => {
    test("インスタンスの作成に失敗する", () => {
      expect(() => new Decimal(NaN))
        .toThrowErrorMatchingSnapshot();
    });
  });
});

test("big.js の静的プロパティーにアクセスできる", () => {
  expect(Decimal.DP).toBeTypeOf("number");
});

test("big.js のインスタンスメソッドにアクセスできる", () => {
  const d = new Decimal("0.1")
    .plus("0.1")
    .plus(new Decimal("0.1"));

  expect(d).toBeInstanceOf(Decimal);
  expect(String(d)).toBe("0.3");
});

test("Decimal であると判定できる", () => {
  expect(isDecimal(new EncodableDecimal("0"))).toBe(true);
  expect(isDecimal(new Decimal(0))).toBe(true);
  expect(isDecimal("0")).toBe(false);
  expect(isDecimal(0)).toBe(false);
});
