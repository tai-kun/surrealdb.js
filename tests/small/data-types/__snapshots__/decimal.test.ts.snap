// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`decode-only/encodable > 無効な文字列 > インスタンスの作成に失敗する 1`] = `
{
  "message": "[big.js] Invalid number",
  "name": "Error",
}
`;

exports[`standard > 無効な数値 > インスタンスの作成に失敗する 1`] = `
{
  "message": "[big.js] Invalid number",
  "name": "Error",
}
`;

exports[`standard > 無効な文字列 > インスタンスの作成に失敗する 1`] = `
{
  "message": "[big.js] Invalid number",
  "name": "Error",
}
`;
